src/Storefront/Controller/AuthController.php

```diff
@@ -358,4 +385,49 @@ public function resetPassword(RequestDataBag $data, SalesChannelContext $context
 
         return $this->redirectToRoute('frontend.account.profile.page');
     }
+
+    /**
+     * @Since("6.0.0.0")
+     * @Route("/account/login/customer/{token}/{salesChannelId}/{customerId}", name="frontend.account.login.customer", methods={"GET"})
+     */
+    public function loginAsCustomer(string $token, string $salesChannelId, string $customerId, SalesChannelContext $context, Request $request): Response
+    {
+        try {
+            $this->loginAsCustomerTokenGenerator->validate($token, $salesChannelId, $customerId);
+
+            $customer = $this->fetchCustomer($customerId, $context->getContext());
+
+            $restoredCart = $this->cartRestorer->restore($customer->getId(), $context);
+
+            $cartToken = $restoredCart->getToken();
+
+            $request->attributes->set(PlatformRequest::ATTRIBUTE_SALES_CHANNEL_CONTEXT_OBJECT, $restoredCart);
+            $request->headers->set(PlatformRequest::HEADER_CONTEXT_TOKEN, $cartToken);
+
+            $event = new CustomerLoginEvent($context, $customer, $cartToken);
+            $this->eventDispatcher->dispatch($event);
+
+            return $this->redirectToRoute('frontend.account.home.page');
+        } catch (InvalidLoginAsCustomerTokenException $exception) {
+            return $this->redirectToRoute('frontend.account.login.page');
+        } catch (CustomerNotFoundByIdException $exception) {
+            return $this->redirectToRoute('frontend.account.login.page');
+        }
+    }
+
+    /**
+     * @throws InconsistentCriteriaIdsException
+     * @throws CustomerNotFoundByIdException
+     */
+    private function fetchCustomer(string $customerId, Context $context): CustomerEntity
+    {
+        /** @var CustomerEntity|null $customer */
+        $customer = $this->customerRepository->search(new Criteria([$customerId]), $context)->get($customerId);
```

Comment:
DAL is not allowed to be called in controllers. Please move the code into a store api route